src/loginServices.js

```diff
@@ -446,36 +452,55 @@ function processSSOLoginResponse(resp) {
   }
 }
 
+/**
+ * handleLoginError
+ * Clear out session data, dispatch authentication errors, then dispatch
+ * okapi-ready to indicate that we're ready to start fresh. Return the
+ * response-body, er, JSON, from the failed auth-n request.
+ *
+ * @param {function} dispatch store.dispatch
+ * @param {Response} resp HTTP response from a failed auth-n request
+ *
+ * @returns {Promise} resolving to the response's JSON
+ */
 function handleLoginError(dispatch, resp) {
-  localforage.removeItem('okapiSess');
-  processBadResponse(dispatch, resp);
-  dispatch(setOkapiReady());
+  return localforage.removeItem('okapiSess')
+    .then(() => processBadResponse(dispatch, resp))
+    .then(responseBody => {
+      dispatch(setOkapiReady());
+      return responseBody;
+    });
 }
 
 /**
  * processOkapiSession
  * create a new okapi session with the response from either a username/password
  * authentication request or a bl-users/_self request.
- * @param {*} okapiUrl
- * @param {*} store
- * @param {*} tenant
- * @param {*} resp
- * @param {*} ssoToken
+ * @param {string} okapiUrl
+ * @param {redux store} store
+ * @param {string} tenant
+ * @param {Response} resp HTTP response
+ * @param {string} ssoToken
+ *
+ * @returns {Promise} resolving with login response body, rejecting with, ummmmm
  */
 function processOkapiSession(okapiUrl, store, tenant, resp, ssoToken) {
   const token = resp.headers.get('X-Okapi-Token') || ssoToken;
   const { dispatch } = store;
 
-  if (resp.status >= 400) {
-    handleLoginError(dispatch, resp);
-  } else {
-    resp.json()
-      .then(json => createOkapiSession(okapiUrl, store, tenant, token, json))
-      .then(() => {
+  if (resp.ok) {
+    return resp.json()
+      .then(json => {
+        return createOkapiSession(okapiUrl, store, tenant, token, json)
+          .then(() => json);
```

Comment:
Nested `then` is hard to read but I get why you had to handle it this way. It looks like you want to get the outer `json` from the original `resp`.

I wonder if the processing of that original response could be handled by the function which calls the `processOkapiSession` so the `processOkapiSession` doesn't really have to worry about processing it (handling errors, etc). The `processOkapiSession` could just get a json as an argument instead of `resp`. With that approach this could be simplified with:

````js
return createOkapiSession(okapiUrl, store, tenant, token, json)
  .then(store.dispatch(setOkapiReady())
  .then(() => json)
````

and actually, if we move the `loadResources` here (from `createOkapiSession`) this could become:

````js
return createOkapiSession(okapiUrl, store, tenant, token, json)
  .then(okapiSess => loadResources(okapiUrl, store, tenant, okapiSess.user.id))
  .then(store.dispatch(setOkapiReady())
  .then(() => json)
````

Reply:
This isn't nitpicky at all; it's a great idea! I'd lean toward a separate PR for it just to avoid the "too many things at once" syndrome, though I'm guilty of that here already. A lot of the work that looks like new work, introducing promise chains and such, isn't really new work though -- at least, it doesn't move anything; it just formalizes/awaits the operations as they already were.